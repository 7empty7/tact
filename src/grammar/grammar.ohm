Tact {

    // Starting point of the program
    Program = ProgramItem*
    ProgramItem = Struct
                | Contract
                | Primitive
                | StaticFunction
                | NativeFunction

    // Built-in declarations
    Primitive = primitive Type ";"

    // Static function
    StaticFunction = Function
    NativeFunction = nameAttribute "(" id ")" native id "(" ListOf<FunctionArg,","> ")" ";" --withVoid
                   | nameAttribute "(" id ")" native id "(" ListOf<FunctionArg,","> ")" ":" Type ";" --withType
    
    // Field declarations
    Type = typeLiteral "?" --optional
         | typeLiteral --required
    Field = var id ":" Type ";"

    // Struct
    Struct = struct id "{" StructBody* "}"
    StructBody = Field

    // Contract
    Contract = contract id "{" ContractBody* "}"
    ContractBody = Field
                 | Function 

    // Function
    FunctionAttribute = public --public
                      | get    --getter
    Function = ListOf<FunctionAttribute," "> fun id "(" ListOf<FunctionArg,","> ")" "{" Statement* "}" --withVoid
             | ListOf<FunctionAttribute," "> fun id "(" ListOf<FunctionArg,","> ")" ":" Type "{" Statement* "}" --withType
             
    FunctionArg = id ":" Type

    // Statements
    Statement = StatementLet
              | StatementBlock
              | StatementReturn
              | StatementCall
              | StatementStaticCall
              | StatementAssign
    StatementBlock = "{" Statement* "}"
    StatementLet = let id ":" Type "=" Expression ";"
    StatementReturn = return Expression ";"
    StatementCall = ExpressionCall ";"
    StatementStaticCall = ExpressionStaticCall ";"
    StatementAssign = LValue Expression ";"

    // L-value
    LValue = id "=" --id
           | id "." LValue --subId

    // Expressions
    Expression = ExpressionOr
    ExpressionOr = ExpressionOr "||" ExpressionAnd --or
                 | ExpressionAnd
    ExpressionAnd = ExpressionAnd "&&" ExpressionCompare --and
                  | ExpressionCompare
    ExpressionCompare = ExpressionCompare "!=" ExpressionAdd --not
                      | ExpressionCompare "==" ExpressionAdd --eq
                      | ExpressionCompare ">" ExpressionAdd --gt
                      | ExpressionCompare ">=" ExpressionAdd --gte
                      | ExpressionCompare "<" ExpressionAdd --lt
                      | ExpressionCompare "<=" ExpressionAdd --lte
                      | ExpressionAdd
    ExpressionAdd = ExpressionAdd "+" ~"+" ExpressionMul --add
                  | ExpressionAdd "-" ~"-" ExpressionMul --sub
                  | ExpressionMul
    ExpressionMul = ExpressionMul "*" ExpressionUnary --mul
                  | ExpressionMul "/" ExpressionUnary --div
                  | ExpressionUnary
    ExpressionUnary = "-" ExpressionUnarySuffix --neg
                    | "+" ExpressionUnarySuffix --add
                    | "!" ExpressionUnarySuffix --not
                    | ExpressionUnarySuffix
    ExpressionUnarySuffix = ExpressionValue "!!" --notNull
                          | ExpressionValue
    ExpressionBracket = "(" Expression ")"

    // Order is important
    ExpressionValue = ExpressionCall
                    | ExpressionField
                    | ExpressionStaticCall
                    | ExpressionBracket
                    | ExpressionNew
                    | integerLiteral
                    | boolLiteral
                    | id
                    
    ExpressionField = ExpressionValue "." id
    ExpressionCall = ExpressionValue "." id "(" ListOf<Expression, ","> ")"
    ExpressionNew = id "{" ListOf<NewParameter, ","> "}"
    NewParameter = id ":" Expression
    ExpressionStaticCall = ExpressionValue "(" ListOf<Expression, ","> ")"

    // Type Literal
    typeLiteral = letterAsciiUC typeLiteralPart*
    typeLiteralPart = letterAscii | digit | "_"

    // Integer Literal
    // hexDigit defined in Ohm's built-in rules (otherwise: hexDigit = "0".."9" | "a".."f" | "A".."F")
    // digit defined in Ohm's built-in rules (otherwise: digit = "0".."9")
    integerLiteral = integerLiteralHex | integerLiteralDec // Order is important
    integerLiteralDec = digit+
    integerLiteralHex = "0x" hexDigit+
                      | "0X" hexDigit+

    // Letters
    letterAsciiLC = "a".."z"
    letterAsciiUC = "A".."Z"
    letterAscii = letterAsciiLC | letterAsciiUC

    // ID Literal
    idStart = letterAscii | "_"
    idPart = letterAscii | digit | "_"
    id = ~reservedWord idStart idPart*

    // Bool Literal
    boolLiteral = ("true" | "false") ~idPart

    // Keywords
    keyword = struct | contract | var | fun | let | return | primitive | extend | native | public | get | null // Order is important
    struct = "struct" ~idPart
    contract = "contract" ~idPart
    var = "var" ~idPart
    let = "let" ~idPart
    fun = "fun" ~idPart
    return = "return" ~idPart
    primitive = "primitive" ~idPart
    extend = "extend" ~idPart
    native = "native" ~idPart
    public = "public" ~idPart
    null = "null" ~idPart
    get = "get" ~idPart

    // Attributes
    nameAttribute = "@name"

    // Reserved
    reservedWord = keyword
}
