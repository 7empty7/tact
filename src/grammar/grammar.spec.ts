import { parse } from "./grammar";
import { ASTRef, __DANGER_resetNodeId } from "../ast/ast";
import { loadCases } from "../utils/loadCases";

expect.addSnapshotSerializer({
    test: (src) => src instanceof ASTRef,
    print: (src) => `${(src as ASTRef).contents}`
});

describe('grammar', () => {
    
    beforeEach(() => {
        __DANGER_resetNodeId();
    });

    for (let r of loadCases(__dirname + "/test/")) {
        it('should parse ' + r.name, () => {
            expect(parse(r.code)).toMatchSnapshot();
        });
    }

    for (let r of loadCases(__dirname + "/test-failed/")) {
        it('should fail ' + r.name, () => {
            expect(() => parse(r.code)).toThrowErrorMatchingSnapshot();
        });
    }
});