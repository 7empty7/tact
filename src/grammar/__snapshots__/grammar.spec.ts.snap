// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grammar should parse case-0.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 4,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return 0;
},
      "return": {
        "id": 1,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 2,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 3,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
  ],
  "id": 5,
  "kind": "program",
}
`;

exports[`grammar should parse case-1.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 15,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return (0 + 1) * 10 / 20;
},
      "return": {
        "id": 6,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 13,
            "kind": "op_binary",
            "left": {
              "id": 11,
              "kind": "op_binary",
              "left": {
                "id": 9,
                "kind": "op_binary",
                "left": {
                  "id": 7,
                  "kind": "number",
                  "ref": 0,
                  "value": 0n,
                },
                "op": "+",
                "ref": 0 + 1,
                "right": {
                  "id": 8,
                  "kind": "number",
                  "ref": 1,
                  "value": 1n,
                },
              },
              "op": "*",
              "ref": (0 + 1) * 10,
              "right": {
                "id": 10,
                "kind": "number",
                "ref": 10,
                "value": 10n,
              },
            },
            "op": "/",
            "ref": (0 + 1) * 10 / 20,
            "right": {
              "id": 12,
              "kind": "number",
              "ref": 20,
              "value": 20n,
            },
          },
          "id": 14,
          "kind": "statement_return",
          "ref": return (0 + 1) * 10 / 20;,
        },
      ],
    },
  ],
  "id": 16,
  "kind": "program",
}
`;

exports[`grammar should parse case-2.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 84,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return (0 + 1) * 10 / 20 != 10;
},
      "return": {
        "id": 73,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 82,
            "kind": "op_binary",
            "left": {
              "id": 80,
              "kind": "op_binary",
              "left": {
                "id": 78,
                "kind": "op_binary",
                "left": {
                  "id": 76,
                  "kind": "op_binary",
                  "left": {
                    "id": 74,
                    "kind": "number",
                    "ref": 0,
                    "value": 0n,
                  },
                  "op": "+",
                  "ref": 0 + 1,
                  "right": {
                    "id": 75,
                    "kind": "number",
                    "ref": 1,
                    "value": 1n,
                  },
                },
                "op": "*",
                "ref": (0 + 1) * 10,
                "right": {
                  "id": 77,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
              },
              "op": "/",
              "ref": (0 + 1) * 10 / 20,
              "right": {
                "id": 79,
                "kind": "number",
                "ref": 20,
                "value": 20n,
              },
            },
            "op": "!=",
            "ref": (0 + 1) * 10 / 20 != 10,
            "right": {
              "id": 81,
              "kind": "number",
              "ref": 10,
              "value": 10n,
            },
          },
          "id": 83,
          "kind": "statement_return",
          "ref": return (0 + 1) * 10 / 20 != 10;,
        },
      ],
    },
  ],
  "id": 85,
  "kind": "program",
}
`;

exports[`grammar should parse case-3.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 99,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return (0 + 1) * 10 / 20 != 10 * someId;
},
      "return": {
        "id": 86,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 97,
            "kind": "op_binary",
            "left": {
              "id": 93,
              "kind": "op_binary",
              "left": {
                "id": 91,
                "kind": "op_binary",
                "left": {
                  "id": 89,
                  "kind": "op_binary",
                  "left": {
                    "id": 87,
                    "kind": "number",
                    "ref": 0,
                    "value": 0n,
                  },
                  "op": "+",
                  "ref": 0 + 1,
                  "right": {
                    "id": 88,
                    "kind": "number",
                    "ref": 1,
                    "value": 1n,
                  },
                },
                "op": "*",
                "ref": (0 + 1) * 10,
                "right": {
                  "id": 90,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
              },
              "op": "/",
              "ref": (0 + 1) * 10 / 20,
              "right": {
                "id": 92,
                "kind": "number",
                "ref": 20,
                "value": 20n,
              },
            },
            "op": "!=",
            "ref": (0 + 1) * 10 / 20 != 10 * someId,
            "right": {
              "id": 96,
              "kind": "op_binary",
              "left": {
                "id": 94,
                "kind": "number",
                "ref": 10,
                "value": 10n,
              },
              "op": "*",
              "ref": 10 * someId,
              "right": {
                "id": 95,
                "kind": "id",
                "ref": someId,
                "value": "someId",
              },
            },
          },
          "id": 98,
          "kind": "statement_return",
          "ref": return (0 + 1) * 10 / 20 != 10 * someId;,
        },
      ],
    },
  ],
  "id": 100,
  "kind": "program",
}
`;

exports[`grammar should parse case-4.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 122,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < 10;
},
      "return": {
        "id": 101,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 120,
            "kind": "op_binary",
            "left": {
              "id": 112,
              "kind": "op_binary",
              "left": {
                "id": 108,
                "kind": "op_binary",
                "left": {
                  "id": 106,
                  "kind": "op_binary",
                  "left": {
                    "id": 104,
                    "kind": "op_binary",
                    "left": {
                      "id": 102,
                      "kind": "number",
                      "ref": 0,
                      "value": 0n,
                    },
                    "op": "+",
                    "ref": 0 + 1,
                    "right": {
                      "id": 103,
                      "kind": "number",
                      "ref": 1,
                      "value": 1n,
                    },
                  },
                  "op": "*",
                  "ref": (0 + 1) * 10,
                  "right": {
                    "id": 105,
                    "kind": "number",
                    "ref": 10,
                    "value": 10n,
                  },
                },
                "op": "/",
                "ref": (0 + 1) * 10 / 20,
                "right": {
                  "id": 107,
                  "kind": "number",
                  "ref": 20,
                  "value": 20n,
                },
              },
              "op": "!=",
              "ref": (0 + 1) * 10 / 20 != 10 * someId,
              "right": {
                "id": 111,
                "kind": "op_binary",
                "left": {
                  "id": 109,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
                "op": "*",
                "ref": 10 * someId,
                "right": {
                  "id": 110,
                  "kind": "id",
                  "ref": someId,
                  "value": "someId",
                },
              },
            },
            "op": "||",
            "ref": (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < 10,
            "right": {
              "id": 119,
              "kind": "op_binary",
              "left": {
                "id": 115,
                "kind": "op_binary",
                "left": {
                  "id": 113,
                  "kind": "id",
                  "ref": some2,
                  "value": "some2",
                },
                "op": ">",
                "ref": some2 > 10,
                "right": {
                  "id": 114,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
              },
              "op": "&&",
              "ref": some2 > 10 && some3 < 10,
              "right": {
                "id": 118,
                "kind": "op_binary",
                "left": {
                  "id": 116,
                  "kind": "id",
                  "ref": some3,
                  "value": "some3",
                },
                "op": ">",
                "ref": some3 < 10,
                "right": {
                  "id": 117,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
              },
            },
          },
          "id": 121,
          "kind": "statement_return",
          "ref": return (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < 10;,
        },
      ],
    },
  ],
  "id": 123,
  "kind": "program",
}
`;

exports[`grammar should parse case-5.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 147,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < 123.add(10);
},
      "return": {
        "id": 124,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 145,
            "kind": "op_binary",
            "left": {
              "id": 135,
              "kind": "op_binary",
              "left": {
                "id": 131,
                "kind": "op_binary",
                "left": {
                  "id": 129,
                  "kind": "op_binary",
                  "left": {
                    "id": 127,
                    "kind": "op_binary",
                    "left": {
                      "id": 125,
                      "kind": "number",
                      "ref": 0,
                      "value": 0n,
                    },
                    "op": "+",
                    "ref": 0 + 1,
                    "right": {
                      "id": 126,
                      "kind": "number",
                      "ref": 1,
                      "value": 1n,
                    },
                  },
                  "op": "*",
                  "ref": (0 + 1) * 10,
                  "right": {
                    "id": 128,
                    "kind": "number",
                    "ref": 10,
                    "value": 10n,
                  },
                },
                "op": "/",
                "ref": (0 + 1) * 10 / 20,
                "right": {
                  "id": 130,
                  "kind": "number",
                  "ref": 20,
                  "value": 20n,
                },
              },
              "op": "!=",
              "ref": (0 + 1) * 10 / 20 != 10 * someId,
              "right": {
                "id": 134,
                "kind": "op_binary",
                "left": {
                  "id": 132,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
                "op": "*",
                "ref": 10 * someId,
                "right": {
                  "id": 133,
                  "kind": "id",
                  "ref": someId,
                  "value": "someId",
                },
              },
            },
            "op": "||",
            "ref": (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < 123.add(10),
            "right": {
              "id": 144,
              "kind": "op_binary",
              "left": {
                "id": 138,
                "kind": "op_binary",
                "left": {
                  "id": 136,
                  "kind": "id",
                  "ref": some2,
                  "value": "some2",
                },
                "op": ">",
                "ref": some2 > 10,
                "right": {
                  "id": 137,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
              },
              "op": "&&",
              "ref": some2 > 10 && some3 < 123.add(10),
              "right": {
                "id": 143,
                "kind": "op_binary",
                "left": {
                  "id": 139,
                  "kind": "id",
                  "ref": some3,
                  "value": "some3",
                },
                "op": ">",
                "ref": some3 < 123.add(10),
                "right": {
                  "args": [
                    {
                      "id": 141,
                      "kind": "number",
                      "ref": 10,
                      "value": 10n,
                    },
                  ],
                  "id": 142,
                  "kind": "op_call",
                  "name": "add",
                  "ref": 123.add(10),
                  "src": {
                    "id": 140,
                    "kind": "number",
                    "ref": 123,
                    "value": 123n,
                  },
                },
              },
            },
          },
          "id": 146,
          "kind": "statement_return",
          "ref": return (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < 123.add(10);,
        },
      ],
    },
  ],
  "id": 148,
  "kind": "program",
}
`;

exports[`grammar should parse case-6.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 173,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < abs(123.add(10));
},
      "return": {
        "id": 149,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 171,
            "kind": "op_binary",
            "left": {
              "id": 160,
              "kind": "op_binary",
              "left": {
                "id": 156,
                "kind": "op_binary",
                "left": {
                  "id": 154,
                  "kind": "op_binary",
                  "left": {
                    "id": 152,
                    "kind": "op_binary",
                    "left": {
                      "id": 150,
                      "kind": "number",
                      "ref": 0,
                      "value": 0n,
                    },
                    "op": "+",
                    "ref": 0 + 1,
                    "right": {
                      "id": 151,
                      "kind": "number",
                      "ref": 1,
                      "value": 1n,
                    },
                  },
                  "op": "*",
                  "ref": (0 + 1) * 10,
                  "right": {
                    "id": 153,
                    "kind": "number",
                    "ref": 10,
                    "value": 10n,
                  },
                },
                "op": "/",
                "ref": (0 + 1) * 10 / 20,
                "right": {
                  "id": 155,
                  "kind": "number",
                  "ref": 20,
                  "value": 20n,
                },
              },
              "op": "!=",
              "ref": (0 + 1) * 10 / 20 != 10 * someId,
              "right": {
                "id": 159,
                "kind": "op_binary",
                "left": {
                  "id": 157,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
                "op": "*",
                "ref": 10 * someId,
                "right": {
                  "id": 158,
                  "kind": "id",
                  "ref": someId,
                  "value": "someId",
                },
              },
            },
            "op": "||",
            "ref": (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < abs(123.add(10)),
            "right": {
              "id": 170,
              "kind": "op_binary",
              "left": {
                "id": 163,
                "kind": "op_binary",
                "left": {
                  "id": 161,
                  "kind": "id",
                  "ref": some2,
                  "value": "some2",
                },
                "op": ">",
                "ref": some2 > 10,
                "right": {
                  "id": 162,
                  "kind": "number",
                  "ref": 10,
                  "value": 10n,
                },
              },
              "op": "&&",
              "ref": some2 > 10 && some3 < abs(123.add(10)),
              "right": {
                "id": 169,
                "kind": "op_binary",
                "left": {
                  "id": 164,
                  "kind": "id",
                  "ref": some3,
                  "value": "some3",
                },
                "op": ">",
                "ref": some3 < abs(123.add(10)),
                "right": {
                  "args": [
                    {
                      "args": [
                        {
                          "id": 166,
                          "kind": "number",
                          "ref": 10,
                          "value": 10n,
                        },
                      ],
                      "id": 167,
                      "kind": "op_call",
                      "name": "add",
                      "ref": 123.add(10),
                      "src": {
                        "id": 165,
                        "kind": "number",
                        "ref": 123,
                        "value": 123n,
                      },
                    },
                  ],
                  "id": 168,
                  "kind": "op_static_call",
                  "name": "abs",
                  "ref": abs(123.add(10)),
                },
              },
            },
          },
          "id": 172,
          "kind": "statement_return",
          "ref": return (0 + 1) * 10 / 20 != 10 * someId || some2 > 10 && some3 < abs(123.add(10));,
        },
      ],
    },
  ],
  "id": 174,
  "kind": "program",
}
`;

exports[`grammar should parse case-7.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 177,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc() {
    return 0;
},
      "return": null,
      "statements": [
        {
          "expression": {
            "id": 175,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 176,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    {
      "args": [],
      "attribute": [],
      "id": 181,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return 0;
},
      "return": {
        "id": 178,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 179,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 180,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    {
      "args": [],
      "attribute": [],
      "id": 185,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Bool {
    return 0;
},
      "return": {
        "id": 182,
        "kind": "type_ref",
        "name": "Bool",
        "optional": false,
        "ref": Bool,
      },
      "statements": [
        {
          "expression": {
            "id": 183,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 184,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
  ],
  "id": 186,
  "kind": "program",
}
`;

exports[`grammar should parse case-8.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "id": 187,
      "kind": "def_native_function",
      "name": "testFunc",
      "nativeName": "native_name_1",
      "ref": @name(native_name_1)
native testFunc();,
      "return": null,
    },
    {
      "args": [],
      "id": 189,
      "kind": "def_native_function",
      "name": "testFunc",
      "nativeName": "native_name_2",
      "ref": @name(native_name_2)
native testFunc(): Int;,
      "return": {
        "id": 188,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
    },
    {
      "args": [],
      "id": 191,
      "kind": "def_native_function",
      "name": "testFunc",
      "nativeName": "native_name_3",
      "ref": @name(native_name_3)
native testFunc(): Bool;,
      "return": {
        "id": 190,
        "kind": "type_ref",
        "name": "Bool",
        "optional": false,
        "ref": Bool,
      },
    },
  ],
  "id": 192,
  "kind": "program",
}
`;

exports[`grammar should parse case-9.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 203,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return 1 + 2 + 3!! > 123;
},
      "return": {
        "id": 193,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 201,
            "kind": "op_binary",
            "left": {
              "id": 199,
              "kind": "op_binary",
              "left": {
                "id": 196,
                "kind": "op_binary",
                "left": {
                  "id": 194,
                  "kind": "number",
                  "ref": 1,
                  "value": 1n,
                },
                "op": "+",
                "ref": 1 + 2,
                "right": {
                  "id": 195,
                  "kind": "number",
                  "ref": 2,
                  "value": 2n,
                },
              },
              "op": "+",
              "ref": 1 + 2 + 3!!,
              "right": {
                "id": 198,
                "kind": "op_unary",
                "op": "!!",
                "ref": 3!!,
                "right": {
                  "id": 197,
                  "kind": "number",
                  "ref": 3,
                  "value": 3n,
                },
              },
            },
            "op": ">",
            "ref": 1 + 2 + 3!! > 123,
            "right": {
              "id": 200,
              "kind": "number",
              "ref": 123,
              "value": 123n,
            },
          },
          "id": 202,
          "kind": "statement_return",
          "ref": return 1 + 2 + 3!! > 123;,
        },
      ],
    },
  ],
  "id": 204,
  "kind": "program",
}
`;

exports[`grammar should parse case-10.tact 1`] = `
{
  "entries": [
    {
      "args": [],
      "attribute": [],
      "id": 29,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return 1 + 2 + (123 + 3)!! > 123;
},
      "return": {
        "id": 17,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 27,
            "kind": "op_binary",
            "left": {
              "id": 25,
              "kind": "op_binary",
              "left": {
                "id": 20,
                "kind": "op_binary",
                "left": {
                  "id": 18,
                  "kind": "number",
                  "ref": 1,
                  "value": 1n,
                },
                "op": "+",
                "ref": 1 + 2,
                "right": {
                  "id": 19,
                  "kind": "number",
                  "ref": 2,
                  "value": 2n,
                },
              },
              "op": "+",
              "ref": 1 + 2 + (123 + 3)!!,
              "right": {
                "id": 24,
                "kind": "op_unary",
                "op": "!!",
                "ref": (123 + 3)!!,
                "right": {
                  "id": 23,
                  "kind": "op_binary",
                  "left": {
                    "id": 21,
                    "kind": "number",
                    "ref": 123,
                    "value": 123n,
                  },
                  "op": "+",
                  "ref": 123 + 3,
                  "right": {
                    "id": 22,
                    "kind": "number",
                    "ref": 3,
                    "value": 3n,
                  },
                },
              },
            },
            "op": ">",
            "ref": 1 + 2 + (123 + 3)!! > 123,
            "right": {
              "id": 26,
              "kind": "number",
              "ref": 123,
              "value": 123n,
            },
          },
          "id": 28,
          "kind": "statement_return",
          "ref": return 1 + 2 + (123 + 3)!! > 123;,
        },
      ],
    },
  ],
  "id": 30,
  "kind": "program",
}
`;

exports[`grammar should parse case-11.tact 1`] = `
{
  "entries": [
    {
      "args": [
        {
          "id": 33,
          "kind": "def_argument",
          "name": "src",
          "ref": src: Int?,
          "type": {
            "id": 32,
            "kind": "type_ref",
            "name": "Int",
            "optional": true,
            "ref": Int?,
          },
        },
      ],
      "attribute": [],
      "id": 45,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(src: Int?): Int {
    return 1 + 2 + (123 + 3)!! > 123;
},
      "return": {
        "id": 31,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 43,
            "kind": "op_binary",
            "left": {
              "id": 41,
              "kind": "op_binary",
              "left": {
                "id": 36,
                "kind": "op_binary",
                "left": {
                  "id": 34,
                  "kind": "number",
                  "ref": 1,
                  "value": 1n,
                },
                "op": "+",
                "ref": 1 + 2,
                "right": {
                  "id": 35,
                  "kind": "number",
                  "ref": 2,
                  "value": 2n,
                },
              },
              "op": "+",
              "ref": 1 + 2 + (123 + 3)!!,
              "right": {
                "id": 40,
                "kind": "op_unary",
                "op": "!!",
                "ref": (123 + 3)!!,
                "right": {
                  "id": 39,
                  "kind": "op_binary",
                  "left": {
                    "id": 37,
                    "kind": "number",
                    "ref": 123,
                    "value": 123n,
                  },
                  "op": "+",
                  "ref": 123 + 3,
                  "right": {
                    "id": 38,
                    "kind": "number",
                    "ref": 3,
                    "value": 3n,
                  },
                },
              },
            },
            "op": ">",
            "ref": 1 + 2 + (123 + 3)!! > 123,
            "right": {
              "id": 42,
              "kind": "number",
              "ref": 123,
              "value": 123n,
            },
          },
          "id": 44,
          "kind": "statement_return",
          "ref": return 1 + 2 + (123 + 3)!! > 123;,
        },
      ],
    },
  ],
  "id": 46,
  "kind": "program",
}
`;

exports[`grammar should parse case-12.tact 1`] = `
{
  "entries": [
    {
      "args": [
        {
          "id": 49,
          "kind": "def_argument",
          "name": "src",
          "ref": src: Int?,
          "type": {
            "id": 48,
            "kind": "type_ref",
            "name": "Int",
            "optional": true,
            "ref": Int?,
          },
        },
      ],
      "attribute": [],
      "id": 58,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(src: Int?): Int {
    if (src != null) {
        return src;
    }
    return 0;
},
      "return": {
        "id": 47,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "elseif": null,
          "expression": {
            "id": 52,
            "kind": "op_binary",
            "left": {
              "id": 50,
              "kind": "id",
              "ref": src,
              "value": "src",
            },
            "op": "!=",
            "ref": src != null,
            "right": {
              "id": 51,
              "kind": "null",
              "ref": null,
            },
          },
          "falseStatements": [],
          "id": 55,
          "kind": "statement_condition",
          "ref": if (src != null) {
        return src;
    },
          "trueStatements": [
            {
              "expression": {
                "id": 53,
                "kind": "id",
                "ref": src,
                "value": "src",
              },
              "id": 54,
              "kind": "statement_return",
              "ref": return src;,
            },
          ],
        },
        {
          "expression": {
            "id": 56,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 57,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
  ],
  "id": 59,
  "kind": "program",
}
`;

exports[`grammar should parse case-13.tact 1`] = `
{
  "entries": [
    {
      "args": [
        {
          "id": 62,
          "kind": "def_argument",
          "name": "src",
          "ref": src: Int?,
          "type": {
            "id": 61,
            "kind": "type_ref",
            "name": "Int",
            "optional": true,
            "ref": Int?,
          },
        },
      ],
      "attribute": [],
      "id": 71,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(src: Int?): Int {
    if (src != null) {
        return src;
    } else {
        return 10;
    }
},
      "return": {
        "id": 60,
        "kind": "type_ref",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "elseif": null,
          "expression": {
            "id": 65,
            "kind": "op_binary",
            "left": {
              "id": 63,
              "kind": "id",
              "ref": src,
              "value": "src",
            },
            "op": "!=",
            "ref": src != null,
            "right": {
              "id": 64,
              "kind": "null",
              "ref": null,
            },
          },
          "falseStatements": [
            {
              "expression": {
                "id": 68,
                "kind": "number",
                "ref": 10,
                "value": 10n,
              },
              "id": 69,
              "kind": "statement_return",
              "ref": return 10;,
            },
          ],
          "id": 70,
          "kind": "statement_condition",
          "ref": if (src != null) {
        return src;
    } else {
        return 10;
    },
          "trueStatements": [
            {
              "expression": {
                "id": 66,
                "kind": "id",
                "ref": src,
                "value": "src",
              },
              "id": 67,
              "kind": "statement_return",
              "ref": return src;,
            },
          ],
        },
      ],
    },
  ],
  "id": 72,
  "kind": "program",
}
`;
